File: src/lib/availability.ts
import { expireOverdue, nowMinutesTripoli } from "./booking";
import { remainingBookingsForServices, remainingOf, staffSnapshots } from "./capacity";
import { db } from "./db";
import { addDaysISO, todayISO, weekdayOf } from "./utils";
import type { LoadedBusiness } from "./workspace";
import { activeMembers, memberCanPerform } from "./workspace";

export async function publicAvailability(
  business: LoadedBusiness,
  args: { date: string; serviceIds: string[] },
) {
  await expireOverdue(db, business.ownerId);
  const fresh = await db.business.findUniqueOrThrow({
    where: { id: business.id },
    include: {
      owner: true,
      members: { include: { services: true, artist: { select: { id: true, name: true, phone: true, slug: true } } } },
      shifts: { orderBy: { sortOrder: "asc" } },
      hours: true,
      blocked: true,
      services: { select: { id: true, kind: true, active: true } },
    },
  });

  const blocked = fresh.blocked.some((row) => row.date === args.date);
  const open = fresh.hours.some((row) => row.dayOfWeek === weekdayOf(args.date));
  const mode = fresh.scheduleMode === "HOURLY" ? "HOURLY" : fresh.scheduleMode === "DAY" ? "DAY" : "SHIFT";
  const minNotice = args.date === todayISO() ? nowMinutesTripoli() + fresh.minNoticeHours * 60 : 0;
  const lastDay = addDaysISO(todayISO(), fresh.bookingHorizonDays - 1);
  const inHorizon = args.date >= todayISO() && args.date <= lastDay;

  if (!inHorizon) {
    return { mode, remainingDay: 0, shifts: [], staff: [], available: false, reason: "HORIZON" as const };
  }
  if (blocked) {
    return { mode, remainingDay: 0, shifts: [], staff: [], available: false, reason: "BLOCKED" as const };
  }
  if (!open) {
    return { mode, remainingDay: 0, shifts: [], staff: [], available: false, reason: "CLOSED" as const };
  }

  const dayStaff = await staffSnapshots(db, fresh, args.date, null);
  const remainingDay = remainingBookingsForServices(fresh, dayStaff, args.serviceIds);

  const shifts = [];
  if (mode === "SHIFT") {
    for (const shift of fresh.shifts.filter((row) => row.active)) {
      if (args.date === todayISO() && shift.endMin <= minNotice) continue;
      const staff = await staffSnapshots(db, fresh, args.date, shift.id);
      const remaining = remainingBookingsForServices(fresh, staff, args.serviceIds);
      shifts.push({
        id: shift.id,
        key: shift.key,
        nameAr: shift.nameAr,
        nameEn: shift.nameEn,
        startMin: shift.startMin,
        endMin: shift.endMin,
        remaining,
        capacity: shift.capacity,
      });
    }
  }

  const serviceKind = (id: string) => fresh.services.find((row) => row.id === id)?.kind;

  const staff = activeMembers(fresh)
    .filter((member) => args.serviceIds.some((id) => memberCanPerform(member, id, serviceKind(id))))
    .map((member) => {
      const row = dayStaff.find((item) => item.id === member.id);
      return {
        id: member.id,
        name: member.name,
        remaining: row ? remainingOf(row) : 0,
        serviceIds: args.serviceIds.filter((id) => memberCanPerform(member, id, serviceKind(id))),
      };
    });

  const available =
    mode === "SHIFT" ? shifts.some((shift) => shift.remaining > 0) : remainingDay > 0;

  return {
    mode,
    remainingDay,
    shifts,
    staff,
    available,
    reason: available ? ("OK" as const) : ("FULL" as const),
  };
}

File: backend/bridey_api/availability.py
from __future__ import annotations

from bridey_api.capacity import member_can_perform, remaining_bookings_for_services, remaining_of, staff_snapshots
from bridey_api.dates import add_days_iso, now_minutes_tripoli, today_iso, weekday_of
from bridey_api.expire import expire_overdue
from bridey_api.workspace import active_members, load_business


def public_availability(session, business, *, date: str, service_ids: list[str]) -> dict:
    expire_overdue(session, business.owner_id)
    fresh = load_business(session, business.id)
    blocked = any(row.date == date for row in fresh.blocked)
    open_day = any(row.day_of_week == weekday_of(date) for row in fresh.hours)
    mode = "HOURLY" if fresh.schedule_mode == "HOURLY" else "DAY" if fresh.schedule_mode == "DAY" else "SHIFT"
    min_notice = now_minutes_tripoli() + fresh.min_notice_hours * 60 if date == today_iso() else 0
    last_day = add_days_iso(today_iso(), fresh.booking_horizon_days - 1)
    in_horizon = today_iso() <= date <= last_day

    empty = {"mode": mode, "remainingDay": 0, "shifts": [], "staff": []}
    if not in_horizon:
        return {**empty, "available": False, "reason": "HORIZON"}
    if blocked:
        return {**empty, "available": False, "reason": "BLOCKED"}
    if not open_day:
        return {**empty, "available": False, "reason": "CLOSED"}

    day_staff = staff_snapshots(session, fresh, date, None)
    remaining_day = remaining_bookings_for_services(fresh, day_staff, service_ids)

    shifts = []
    if mode == "SHIFT":
        for shift in [row for row in fresh.shifts if row.active]:
            if date == today_iso() and shift.end_min <= min_notice:
                continue
            staff = staff_snapshots(session, fresh, date, shift.id)
            remaining = remaining_bookings_for_services(fresh, staff, service_ids)
            shifts.append(
                {
                    "id": shift.id,
                    "key": shift.key,
                    "nameAr": shift.name_ar,
                    "nameEn": shift.name_en,
                    "startMin": shift.start_min,
                    "endMin": shift.end_min,
                    "remaining": remaining,
                    "capacity": shift.capacity,
                }
            )

    def service_kind(service_id: str) -> str | None:
        service = next((row for row in fresh.services if row.id == service_id), None)
        return service.kind if service else None

    staff = []
    for member in active_members(fresh):
        if not any(member_can_perform(member, service_id, service_kind(service_id)) for service_id in service_ids):
            continue
        row = next((item for item in day_staff if item["id"] == member.id), None)
        staff.append(
            {
                "id": member.id,
                "name": member.name,
                "remaining": remaining_of(row) if row else 0,
                "serviceIds": [service_id for service_id in service_ids if member_can_perform(member, service_id, service_kind(service_id))],
            }
        )

    available = any(shift["remaining"] > 0 for shift in shifts) if mode == "SHIFT" else remaining_day > 0
    return {
        "mode": mode,
        "remainingDay": remaining_day,
        "shifts": shifts,
        "staff": staff,
        "available": available,
        "reason": "OK" if available else "FULL",
    }
